react/features/authentication/functions.web.ts

```diff
@@ -0,0 +1,80 @@
+import base64js from 'base64-js';
+
+import { IConfig } from '../base/config/configType';
+import { browser } from '../base/lib-jitsi-meet';
+
+import { _getTokenAuthState } from './functions.any';
+
+export * from './functions.any';
+
+/**
+ * Based on rfc7636 we need a random string for a code verifier.
+ */
+const POSSIBLE_CHARS = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
+
+/**
+ * Creates the URL pointing to JWT token authentication service. It is
+ * formatted from the 'urlPattern' argument which can contain the following
+ * constants:
+ * '{room}' - name of the conference room passed as <tt>roomName</tt>
+ * argument to this method.
+ *
+ * @param {Object} config - Configuration state object from store. A URL pattern pointing to the login service.
+ * @param {string} roomName - The name of the conference room for which the user will be authenticated.
+ * @param {string} tenant - The name of the conference tenant.
+ * @param {string} skipPrejoin - The name of the conference room for which the user will be authenticated.
+ *
+ * @returns {Promise<string|undefined>} - The URL pointing to JWT login service or
+ * <tt>undefined</tt> if the pattern stored in config is not a string and the URL can not be
+ * constructed.
+ */
+export const getTokenAuthUrl = (
+        config: IConfig,
+        roomName: string | undefined,
+        tenant: string | undefined,
+        skipPrejoin: boolean | undefined = false): Promise<string | undefined> => {
+
+    let url = config.tokenAuthUrl;
+
+    if (!url || !roomName) {
+        return Promise.resolve(undefined);
+    }
+
+    if (url.indexOf('{state}')) {
+        const state = _getTokenAuthState(roomName, tenant, skipPrejoin);
+
+        if (browser.isElectron()) {
+            // @ts-ignore
+            state.electron = true;
+        }
+
+        url = url.replace('{state}', encodeURIComponent(JSON.stringify(state)));
+    }
+
+    url = url.replace('{room}', roomName);
+
+    if (url.indexOf('{code_challenge}')) {
+        let codeVerifier = '';
+
+        // random string
+        for (let i = 0; i < 64; i++) {
+            codeVerifier += POSSIBLE_CHARS.charAt(Math.floor(Math.random() * POSSIBLE_CHARS.length));
+        }
+
+        window.sessionStorage.setItem('code_verifier', codeVerifier);
+
+        return window.crypto.subtle.digest('SHA-256', new TextEncoder().encode(codeVerifier))
```

Comment:
Should the digest type be configurable with another parameter?

Reply:
This is the common one, I see. There is a plain option that can be used for compatibility, but its not recommended

Reply:
https://datatracker.ietf.org/doc/html/rfc7636#section-4.2

The rfc says 
  If the client is capable of using "S256", it MUST use "S256", as
   "S256" is Mandatory To Implement (MTI) on the server.

Reply:
Disregard my commend then!